ngraph/test/type_prop/matmul.cpp

```diff
@@ -108,3 +108,40 @@ TEST(type_prop, matmul_4D_transpose_b)
     ASSERT_EQ(matmul->get_element_type(), element::f32);
     ASSERT_EQ(matmul->get_shape(), (Shape{2, 2, 3, 4}));
 }
+
+TEST(type_prop, matmul_dynamic_5D_transpose_b)
+{
+    Dimension dynamic = Dimension::dynamic();
+    auto A =
+        make_shared<op::Parameter>(element::f32, PartialShape{dynamic, 4, dynamic, dynamic, 6});
+    auto B = make_shared<op::Parameter>(element::f32, PartialShape{1, dynamic, dynamic, 4, 6});
+
+    auto matmul = make_shared<op::MatMul>(A, B, 0, 1);
+
+    ASSERT_EQ(matmul->get_element_type(), element::f32);
+    ASSERT_EQ(matmul->get_output_partial_shape(0), (PartialShape{dynamic, 4, dynamic, dynamic, 4}));
+}
+
+TEST(type_prop, matmul_dynamic_2D_transpose_a)
+{
+    Dimension dynamic = Dimension::dynamic();
+    auto A = make_shared<op::Parameter>(element::f32, PartialShape{dynamic, 3});
+    auto B = make_shared<op::Parameter>(element::f32, PartialShape{4, dynamic});
+
+    auto matmul = make_shared<op::MatMul>(A, B, 1, 0);
+
+    ASSERT_EQ(matmul->get_element_type(), element::f32);
+    ASSERT_EQ(matmul->get_output_partial_shape(0), (PartialShape{3, dynamic}));
+}
+
+TEST(type_prop, matmul_dynamic_1D_3D)
+{
+    Dimension dynamic = Dimension::dynamic();
+    auto A = make_shared<op::Parameter>(element::f32, PartialShape{dynamic});
+    auto B = make_shared<op::Parameter>(element::f32, PartialShape{2, 4, dynamic});
+
+    auto matmul = make_shared<op::MatMul>(A, B);
+
+    ASSERT_EQ(matmul->get_element_type(), element::f32);
+    ASSERT_EQ(matmul->get_output_partial_shape(0), (PartialShape{2, dynamic}));
```

Comment:
Shouldn't it be
```
A{-1} x B{2, 4, -1} == A{1, -1} x B{2, 4, -1} == A{2, 1, -1} x B{2, 4, -1} == C{2, 1, -1}
```

Reply:
No, here is an example of multiplication 1-D vector with 3-D matrix. It can be disassembled to:
a) A{-1} x B{4, -1}, for vector multiplication it is expected to get also vector on the output with dimension taken from B[0], here {-1};
b) broadcasting batch dimension to output shape.
That's why final output shape is {2, -1}.

Reply:
Matmul shape for vector input is corrected according to spec.